pandas/core/indexes/base.py

```diff
@@ -177,10 +181,18 @@ class Index(IndexOpsMixin, PandasObject):
     str = accessor.AccessorProperty(strings.StringMethods)
 
     def __new__(cls, data=None, dtype=None, copy=False, name=None,
-                fastpath=False, tupleize_cols=True, **kwargs):
+                fastpath=False, tupleize_cols=True, names=None,
+                **kwargs):
 
-        if name is None and hasattr(data, 'name'):
-            name = data.name
+        if names is not None and name is not None:
+            raise TypeError("Can only provide one of `names` and `name`")
+
+        if name is None:
+            if hasattr(data, 'name'):
+                name = data.name
+            # extract `name` from `names` in case MultiIndex cannot be created
+            elif names:
+                name = names[0] if is_list_like(names) else names
```

Comment:
I don't think we really want this ``else names``. Supporting ``names`` mainly makes sense for compatibility with the multi-level case, where it _must_ be iterable anyway.

Reply:
... moreover, it would be nice to check that if we don't create a ``MultiIndex`` and ``names`` is set, ``len(names) == 1``.

Reply:
Well, I totally agree with the first one.
But as for the last comment: I don't quite understand how we can determine in advance that we don't create a `MultiIndex` as the constructor tries to create it nearly at the end and `tupleize_cols` is always True by default. But for the use-case from the issue we need it at the beginning.

Reply:
Or maybe it'd be better to get rid of `names` handling in `__new__` and use `set_names` (that already does all that checks) after an object has been created.